test(sns): Porting sns-testing to the ICP mono repo, Part III (#4705)
dfinity/ic: rs/sns/testing/README.md

targets before starting the scenario:
```
bazel build //rs/pocket_ic_server:pocket-ic-server

> bazel build //rs/sns/testing/...

Author: Let's have a separate section in the README for how to build the required binaries, 

then just write the instructions in terms of the CLI, not in terms of `bazel run ...`

Wdyt?

Reviewer: It does make sense to me if `bazel` allows to run a shell in which CLI binaries are directly available

Reviewer: Do you think it makes sense to add a bazel script that will install all required binaries somewhere in `PATH` (e.g. `/usr/local/bin`)?
So that users will have to run:
```
bazel run //rs/sns/testing:install
```
This will symlink `//rs/sns/testing:sns-testing-init`, `//rs/sns/testing:sns-testing`, `//rs/sns/cli:sns` binaries to `/usr/local/bin/`

Reviewer: Or alternatively, add a shell script to be used as `. scripts/setup.sh`.
This script will build all binaries and add them to `PATH`

Reviewer: Something like
```
#! /usr/bin/env bash

set -e

bazel_targets=(
    "//rs/sns/testing:sns-testing-init"
    "//rs/sns/testing:sns-testing"
    "//rs/sns/cli:sns"
    "//rs/pocket_ic_server:pocket-ic-server"
)

bazel_root="$(bazel info workspace)"
bin_dir="$PWD/.sns-testing-bin"
mkdir -p "$bin_dir"

for target in "${bazel_targets[@]}"; do
    bazel build "$target"
    bazel_binary_path="$(bazel cquery "$target" \
        --output=starlark --starlark:expr="target.files.to_list()[0].path")"
    ln -sf "$bazel_root/$bazel_binary_path" "$bin_dir/"
done

export PATH="$bin_dir:$PATH"
```

Reviewer: The only tricky part is running `sns-testing-init` outside of `bazel run` since it requires env variables with paths to canister WASM modules 😕

Author: Closing for now (discussed over slack), might get back to this part in a bit.

   ```
2) Bootstrap the NNS on the launched PocketIC instance:
   ```

>    bazel run //rs/sns/testing:sns-testing-init -- --server-url "http://127.0.0.1:8888" --state-dir "$PWD/sns-testing" --dev-identity sns-testing --deciding-nns-neuron-id 1

Author: ```suggestion
   bazel run //rs/sns/testing:sns-testing-init -- \
       --server-url "http://127.0.0.1:8888" \
       --state-dir "$PWD/sns-testing" \
       --dev-identity sns-testing \
       --deciding-nns-neuron-id 1
```

   ```
4) Launch the basic SNS testing scenario:
   ```
   bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 run-basic-scenario \
      --dev-identity sns-testing --nns-neuron-id 1 \
      --test-canister-id "$(dfx canister --network http://127.0.0.1:8080 id test)"
   ```

To start the scenario from scratch, you'll need to stop the running `pocket-ic-server` instance and
remove `$PWD/sns-testing` and `$PWD/.dfx` directories before doing the steps mentioned above.

## Deploying user-provided SNS

### Bootstrap the network

1) Launch PocketIC server:
   ```
   bazel run //rs/pocket_ic_server:pocket-ic-server -- --ttl 6000 --port 8888
   ```

2) Bootstrap the NNS on the launched PocketIC instance:
   ```

>    bazel run //rs/sns/testing:sns-testing-init -- --server-url "http://127.0.0.1:8888" --state-dir "$PWD/sns-testing" --dev-identity sns-testing --deciding-nns-neuron-id 1

Author: ```suggestion
   bazel run //rs/sns/testing:sns-testing-init -- \
       --server-url "http://127.0.0.1:8888" \
       --state-dir "$PWD/sns-testing" \
       --dev-identity sns-testing \
       --deciding-nns-neuron-id 1
```

   ```

Once these steps are completed, the PocketIC will expose the IC network HTTP endpoint on "http://127.0.0.1:8080".
`sns-testing` identity will be added as a hotkey to the NNS neuron with the majority voting power, so all NNS proposals made

> by `sns-testing` will be automatically accepted.

Author: Uber nit

```suggestion
by `sns-testing` will be automatically adopted.
```

> Additionally, `nns-init` will output the ID of the NNS neuron that should be used to create NNS proposals.

Author: What's nns-init?

```
...

> Use the following Neuron ID for further testing: 1

Author: ```suggestion
Use the following NNS neuron ID for further testing: 1
```

```

### Create the new SNS

Now you can create a new SNS via NNS proposal.

The suggested way to do this is to use `dfx sns propose` command. For more information please refer to the [documentation](https://internetcomputer.org/docs/building-apps/governing-apps/launching/launch-steps-1proposal#3-submit-nns-proposal-to-create-sns).

> Make sure to use `sns-testing` identity when creating the proposal.

Author: ```suggestion
Make sure to use `sns-testing` identity when creating the proposal, so that it gets instantly adopted.
```

<details>
<summary>Sample SNS proposal creation workflow</summary>
<br>

The example will use `//rs/sns/testing:sns_testing_canister` canister as SNS-controlled canister and will base on the [init YAML file from SNS CLI](../cli/test_sns_init_v2.yaml).

> **Make sure to use `start_time: null` in the swap parameters to ensure that the swap starts right away after the NNS proposal is executed.**

Author: ```suggestion
**While using a custom sns_init.yaml file, make sure to set `start_time: null` in the swap parameters to ensure that the swap starts right away after the NNS proposal is executed.**
```

1) Build and deploy `test` canister:
   ```
   bazel build //rs/sns/testing:sns_testing_canister

>    # 'r7inp-6aaaa-aaaaa-aaabq-cai' is the NNS Root canister

Author: Nit

```
NNS_ROOT="r7inp-6aaaa-aaaaa-aaabq-cai"
```

>    dfx canister --network http://127.0.0.1:8080 --identity sns-testing create test --controller r7inp-6aaaa-aaaaa-aaabq-cai --controller sns-testing --no-wallet

Author: ```suggestion
   dfx canister --network http://127.0.0.1:8080 --identity sns-testing create test --controller $NNS_ROOT --controller sns-testing --no-wallet
```

>    dfx canister --network http://127.0.0.1:8080 --identity sns-testing install test --wasm "$(bazel info bazel-bin)/rs/sns/testing/sns_testing_canister.wasm.gz"

Author: ```suggestion
   dfx canister \
        --network http://127.0.0.1:8080 \
        --identity sns-testing \
        install test \
        --wasm "$(bazel info bazel-bin)/rs/sns/testing/sns_testing_canister.wasm.gz"
```

   ```

2) Adjust init YAML file (you will need [`yq`](https://github.com/mikefarah/yq) to be installed to do this):
   ```

>    yq -i ".dapp_canisters |= [\""$(dfx canister --network http://127.0.0.1:8080 id test)"\"]" ../cli/test_sns_init_v2.yaml

Author: I would not work with the yaml file from another crate, as it adds complexity to the instructions. 

Let's just copy the `../cli/test_sns_init_v2.yaml` file in the beginning of the instructions: 

```
cp ../cli/test_sns_init_v2.yaml sns_init.yaml
```

, explaining that the user may have their own custom config (with some recommendations for testing).

And then the instructions be written against a local file.

   yq -i ".Distribution.Neurons[0].principal |= \""$(dfx identity get-principal --identity sns-testing)"\"" ../cli/test_sns_init_v2.yaml
   yq -i ".Swap.start_time |= null" ../cli/test_sns_init_v2.yaml
   ```
3) Propose to create the new SNS:
   ```
   pushd ../cli
   # //rs/sns/cli:sns doesn't support CLI-provided identities despite '--identity' option
   dfx identity use sns-testing
   bazel run //rs/sns/cli:sns -- propose --network http://127.0.0.1:8080 --neuron-id 1 $PWD/test_sns_init_v2.yaml
   popd
   ```

4) Complete the swap for the newly created SNS
   ```
   SNS_NAME="$(yq -r .name ../cli/test_sns_init_v2.yaml)"
   bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 swap-complete --sns-name "$SNS_NAME"
   ```
</details>

> Once the NNS proposal to create the new SNS is adopted and executed, the SNS swap is supposed to open.

Author: This should probably go between steps 3 and 4

Reviewer: The steps above are hidden in "Sample SNS proposal creation workflow" dropdown. This dropdown showcases an example of SNS creation based on `test_sns_init_v2.yaml`.

While this line introduces the `swap-complete` subcommand interface, which should be agnostic to the specific SNS unlike the instructions above

Use `bazel run //rs/sns/testing:cli -- run swap-complete` to generate swap participations and complete the swap:
```
SNS_NAME="$(yq -r .name <Path to SNS YAML>)"

> bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 swap-complete --sns-name "$SNS_NAME" --follow-principal-neurons "$(dfx identity get-principal --identity sns-testing)"

Author: ```suggestion
bazel run //rs/sns/testing:sns-testing -- \
    --network http://127.0.0.1:8080 \
    swap-complete \
    --sns-name "$SNS_NAME" \
    --follow-principal-neurons "$(dfx identity get-principal --identity sns-testing)"
```

```

This command will generate required number of participations with the sufficient amount of direct participants to complete the swap.

This command has optional `--follow-principal-neurons` and `--follow-neuron` arguments that accept `PrincipalId` and `NeuronId` to make swap
participant neurons follow given neurons.

### SNS voting

> The newly created SNS is now fully functional and is ready for proposals.

Author: Actually, it's also possible to send some proposals during the swap, not only after it completes

Reviewer: Oh, I didn't know that 😅

> The suggested way to submit SNS proposals is to use [`quill`](https://github.com/dfinity/quill/), for more info please refer to the [official documentation](https://internetcomputer.org/docs/building-apps/governing-apps/managing/making-proposals/).

Author: Let's instead recommend that SNS proposals are submitted via `DFX`:

E.g.,

```
dfx --identity sns-testing canister --network ic call $SNS_GOVERNANCE manage_neuron
# Fill out the details for the make-proposal meuron management action in the interactive prompt
```

Reviewer: I think using a `dfx canister call` might be pretty non-trivial when it comes to the `manage_neuron` method since this method requires providing a byte-encoded WASM module and candid arg

Author: Note that if the user types `dfx canister call $SNS_GOVERNANCE` and hits enter, the tool will provide an interactive prompt that supports hex encoded blobs, see screenshot. 

![Screenshot 2025-04-14 at 14 53 59](https://github.com/user-attachments/assets/d4211534-dc55-4fbd-af2c-88b71105f6c7)

NCR for this thread due to prior agreements.

Author: Closing this as we agreed that keeping quill in the readme is fine for now.

   ```

   This command will return the ID of the newly created proposal
   ```
   ...
   Successfully created new proposal with ID 1
   ```

</details>

Once the SNS proposal is created, the voting begins.
To upvote the proposal using Neurons controlled by identities that participated in the swap on the previous step run:
```
bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 sns-proposal-upvote --sns-name "<SNS name>" --proposal-id "<Proposal ID>" --wait
```

## Get ICP tokens

You can use `bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 transfer-icp` to get ICP tokens to your account.

This subcommand supports transfer to the direct account ID, e.g.:
```
bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 transfer-icp --amount 650.0 --to 5c9b28f3e2218975ea76449cf9b949a860d741a3af7dd4dd062f7481e3a99cde
```

Or to the principal by its ID (and optionally subaccount):
```
bazel run //rs/sns/testing:sns-testing -- --network http://127.0.0.1:8080 transfer-icp --amount 650.0 --to-principal fomoo-4i5fe-epl6n-dc7hi-lfqhi-4ii4q-txsav-xvenw-ffacw-youhl-mae
```

> Obtained ICP tokens can be used to participate in SNS swap or you may stake them to get NNS neuron. This can be done either using `quill` or NNS dapp web UI.

Author: Also, ICP is needed to buy cycles that are required for some SNS operations, e.g., upgrading dapps with large Wasms
